File: src/main/kotlin/tables/TemplateVariables.kt
package tables

import org.jetbrains.exposed.v1.core.Table
import org.jetbrains.exposed.v1.datetime.datetime
import org.jetbrains.exposed.v1.datetime.date
import schema.type.TemplateVariableType

// Base table for all template variables
object TemplateVariableBase : Table("template_variable_base") {
    val id = integer("id").autoIncrement()
    val templateId = integer("template_id").references(Templates.id)
    val name = varchar("name", 255)
    val description = text("description").nullable()
    val type = enumerationByName("type", 50, TemplateVariableType::class)
    val required = bool("required").default(false)
    val order = integer("order")
    val createdAt = datetime("created_at")
    val updatedAt = datetime("updated_at")
    override val primaryKey = PrimaryKey(id)

    init {
        uniqueIndex("template_variable_base_name_unique", templateId, name)
    }
}

// Text variable properties table
object TextTemplateVariables : Table("text_template_variables") {
    val id = integer("id").references(TemplateVariableBase.id)
    val minLength = integer("min_length").nullable()
    val maxLength = integer("max_length").nullable()
    val pattern = varchar("pattern", 255).nullable()
    val previewValue = varchar("preview_value", 255).nullable()
    override val primaryKey = PrimaryKey(id)
}

// Number variable properties table
object NumberTemplateVariables : Table("number_template_variables") {
    val id = integer("id").references(TemplateVariableBase.id)
    val minValue = decimal("min_value", 65, 10).nullable()
    val maxValue = decimal("max_value", 65, 10).nullable()
    val decimalPlaces = integer("decimal_places").nullable()
    val previewValue = decimal("preview_value", 65, 10).nullable()
    override val primaryKey = PrimaryKey(id)
}

// Date variable properties table
object DateTemplateVariables : Table("date_template_variables") {
    val id = integer("id").references(TemplateVariableBase.id)
    val minDate = date("min_date").nullable()
    val maxDate = date("max_date").nullable()
    val format = varchar("format", 50).nullable()
    val previewValue = date("preview_value").nullable()
    override val primaryKey = PrimaryKey(id)
}

// Select variable properties table
object SelectTemplateVariables : Table("select_template_variables") {
    val id = integer("id").references(TemplateVariableBase.id)
    val options = text("options").nullable() // JSON array
    val multiple = bool("multiple").nullable()
    val previewValue = varchar("preview_value", 255).nullable()
    override val primaryKey = PrimaryKey(id)
}

File: db/schema/templateVariables.ts
import {
    pgTable,
    serial,
    integer,
    varchar,
    text,
    boolean,
    timestamp,
    decimal,
    json,
    pgEnum,
    uniqueIndex,
} from "drizzle-orm/pg-core";
import { relations } from "drizzle-orm/_relations";
import { templates } from ".";

export const templateVariableTypeEnum = pgEnum("TemplateVariableType", [
    "TEXT",
    "NUMBER",
    "DATE",
    "SELECT",
]);

export const templateVariableBases = pgTable(
    "TemplateVariableBase",
    {
        id: serial("id").primaryKey(),
        templateId: integer("templateId").notNull(),
        name: varchar("name", { length: 255 }).notNull(),
        description: text("description"),
        type: templateVariableTypeEnum("type").notNull(),
        required: boolean("required").notNull().default(false),
        order: integer("order").notNull(),
        createdAt: timestamp("createdAt", { precision: 3 }).notNull(),
        updatedAt: timestamp("updatedAt", { precision: 3 }).notNull(),
    },
    (table) => [
        uniqueIndex("TemplateVariableBase_templateId_name_key").on(
            table.templateId,
            table.name,
        ),
    ],
);

export const textTemplateVariables = pgTable("TextTemplateVariable", {
    id: integer("id").primaryKey(),
    minLength: integer("minLength"),
    maxLength: integer("maxLength"),
    pattern: varchar("pattern", { length: 255 }),
    previewValue: varchar("previewValue", { length: 255 }),
});

export const numberTemplateVariables = pgTable("NumberTemplateVariable", {
    id: integer("id").primaryKey(),
    minValue: decimal("minValue"),
    maxValue: decimal("maxValue"),
    decimalPlaces: integer("decimalPlaces"),
    previewValue: decimal("previewValue"),
});

export const dateTemplateVariables = pgTable("DateTemplateVariable", {
    id: integer("id").primaryKey(),
    minDate: timestamp("minDate", { precision: 3 }),
    maxDate: timestamp("maxDate", { precision: 3 }),
    format: varchar("format", { length: 50 }),
    previewValue: timestamp("previewValue", { precision: 3 }),
});

export const selectTemplateVariables = pgTable("SelectTemplateVariable", {
    id: integer("id").primaryKey(),
    options: json("options"),
    multiple: boolean("multiple"),
    previewValue: varchar("previewValue", { length: 255 }),
});

export const templateVariableBasesRelations = relations(
    templateVariableBases,
    ({ one }) => ({
        template: one(templates, {
            fields: [templateVariableBases.templateId],
            references: [templates.id],
        }),
        textTemplateVariable: one(textTemplateVariables, {
            fields: [templateVariableBases.id],
            references: [textTemplateVariables.id],
        }),
        numberTemplateVariable: one(numberTemplateVariables, {
            fields: [templateVariableBases.id],
            references: [numberTemplateVariables.id],
        }),
        dateTemplateVariable: one(dateTemplateVariables, {
            fields: [templateVariableBases.id],
            references: [dateTemplateVariables.id],
        }),
        selectTemplateVariable: one(selectTemplateVariables, {
            fields: [templateVariableBases.id],
            references: [selectTemplateVariables.id],
        }),
    }),
);
